src/librustc_mir/util/borrowck_errors.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright 2012-2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use rustc::ty::{TyCtxt};
+use rustc_errors::DiagnosticBuilder;
+use syntax_pos::{MultiSpan, Span};
+
+use std::fmt;
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub enum Origin { Ast, Mir }
+
+impl fmt::Display for Origin {
```

Comment:
Hmm apparently I never pushed up the changes to render `Origin::Ast` as empty string. Will address pronto. (My plan had been to have the empty string rendering be conditional, so that when one passes `-Z borrowck-mir`, you get a symmetry between the error messages in terms of whether they say `"... (Ast)"` or `"... (Mir)"`.)